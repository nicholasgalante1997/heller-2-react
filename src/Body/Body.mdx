import { Canvas, Meta } from '@storybook/blocks';
import * as Stories from './Body.stories';

<Meta of={Stories} title="react/components/typography/Body/Docs" />

# Typography - Body

The `<Body />` component is a `Typography` component that's intended for use in paragraphs, bodies of block text, labels, and accent text.
The `Body` typography can be implemented as an `HTMLParagraph` tag, `HTMLSpan` tag, or `HTMLLabel` tag.

<Canvas of={Stories.BodyMain} />
