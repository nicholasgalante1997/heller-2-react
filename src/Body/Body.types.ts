import { HTMLProps } from 'react';
import { HasChildNodes } from '@/types/HasChildren';

type TextHtmlProps = HTMLProps<HTMLSpanElement | HTMLParagraphElement | HTMLLabelElement>;
type InternalBodyTextProps = {
  bold?: boolean;
  italic?: boolean;
  accent?: boolean;
  as: 'p' | 'label' | 'span';
};

export type BodyProps = TextHtmlProps & InternalBodyTextProps & HasChildNodes;
