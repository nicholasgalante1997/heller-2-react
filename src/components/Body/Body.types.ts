import { type HTMLProps } from 'react';
import { type HasChildNodes } from '@/types/HasChildren';

type TextHtmlProps = HTMLProps<HTMLSpanElement | HTMLParagraphElement | HTMLLabelElement>;
interface InternalBodyTextProps {
  bold?: boolean;
  italic?: boolean;
  accent?: boolean;
  as: 'p' | 'label' | 'span';
}

export type BodyProps = TextHtmlProps & InternalBodyTextProps & HasChildNodes;
