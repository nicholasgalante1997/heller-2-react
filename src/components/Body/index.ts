export { Body } from './Body';
export { type BodyProps } from './Body.types';
