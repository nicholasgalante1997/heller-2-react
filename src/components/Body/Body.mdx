import { Canvas, Meta } from '@storybook/blocks';
import * as Stories from './Body.stories';
import { Body } from './Body';

<Meta of={Stories} title="Typography/Body/Docs" />

# Typography - Body

The `<Body />` component is a `Typography` component that's intended for use in paragraphs, bodies of block text, labels, and accent text.
The `Body` typography can be implemented as an `HTMLParagraph` tag, `HTMLSpan` tag, or `HTMLLabel` tag.

<Canvas of={Stories.BodyMain} />
