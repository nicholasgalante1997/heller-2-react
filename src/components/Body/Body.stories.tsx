import React from 'react';
import { type Meta, type StoryObj } from '@storybook/react';
import { Body } from './Body';

const meta: Meta<typeof Body> = {
  title: 'REACT/Components/Typography/Body',
  component: Body
};

export default meta;

type BodyStory = StoryObj<typeof Body>;

const content_alpha = "That's a load balancing candy cane,";
const content_beta = 'you clumsy oaf!';
const author = '- The Simpsons';

export const BodyMain: BodyStory = {
  render: (args) => (
    <Body {...args} as="p">
      {content_alpha}{' '}
      <Body as="span" italic bold>
        {content_beta}
      </Body>{' '}
      {author}
    </Body>
  )
};
