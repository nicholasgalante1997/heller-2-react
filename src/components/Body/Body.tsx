import React, { useMemo } from 'react';
import { type BodyProps } from './Body.types';
import classNames from 'classnames';

const bodyTextMap = new Map<BodyProps['as'], (props: any) => React.JSX.Element>([
  ['p', ({ children, ...rest }) => <p {...rest}>{children}</p>],
  ['span', ({ children, ...rest }) => <span {...rest}>{children}</span>],
  ['label', ({ children, ...rest }) => <label {...rest}>{children}</label>]
]);

function BodyComponent({
  children,
  as,
  italic,
  bold,
  accent,
  className: classNameProp = '',
  ...rest
}: BodyProps) {
  const Component = useMemo(() => bodyTextMap.get(as), [as]);
  if (!Component) {
    console.error('HELLER-2-REACT:::TYPOGRAPHY:::UNSUPPLIED_AS_PROP');
    return false;
  }
  const className = useMemo(
    () =>
      classNames({
        'h-body': true,
        'h-body-bold': bold,
        'h-body-italic': italic,
        'h-body-accent': accent,
        [classNameProp]: !!classNameProp
      }),
    [italic, bold, classNameProp, accent]
  );
  return (
    <Component {...rest} className={className}>
      {children}
    </Component>
  );
}

export const Body = React.memo(BodyComponent);
