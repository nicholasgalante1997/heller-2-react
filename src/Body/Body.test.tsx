import React from 'react';
import { render } from '@testing-library/react';
import { Body } from './Body';

describe('Body', () => {
  test('Snapshot', () => {
    const { container } = render(<Body as="p">Shmallas</Body>);
    expect(container).toMatchSnapshot();
  });
});
