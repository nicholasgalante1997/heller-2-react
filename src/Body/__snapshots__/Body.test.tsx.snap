// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body Snapshot 1`] = `
<div>
  <p
    class="h-body"
  >
    Shmallas
  </p>
</div>
`;
